api/channel.go

```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	w.Write([]byte(channels.ToJson()))
 }
+
+func viewChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+	view := model.ChannelViewFromJson(r.Body)
+
+	go func() {
+		if err := SetActiveChannel(c.Session.UserId, view.ChannelId); err != nil {
+			l4g.Error(err.Error())
+		}
+	}()
+
+	if len(view.ChannelId) > 0 {
+
+		if view.Time == 0 {
+			if result := <-Srv.Store.Channel().UpdateLastViewedAt(view.ChannelId, c.Session.UserId); result.Err != nil {
+				c.Err = result.Err
+				return
+			}
+		} else {
+			if result := <-Srv.Store.Channel().SetLastViewedAt(view.ChannelId, c.Session.UserId, view.Time); result.Err != nil {
+				c.Err = result.Err
+				return
+			}
+		}
+
+		if len(view.PrevChannelId) > 0 {
+			Srv.Store.Channel().UpdateLastViewedAt(view.PrevChannelId, c.Session.UserId)
+
+			// Only clear push notifications if a channel switch occured
+			if *utils.Cfg.EmailSettings.SendPushNotifications && !c.Session.IsMobileApp() {
+				go func() {
+					if result := <-Srv.Store.User().GetUnreadCountForChannel(c.Session.UserId, view.ChannelId); result.Err != nil {
+						l4g.Error(utils.T("api.channel.update_last_viewed_at.get_unread_count_for_channel.error"), c.Session.UserId, view.ChannelId, result.Err.Error())
+					} else {
+						if result.Data.(int64) > 0 {
+							clearPushNotification(c.Session.UserId, view.ChannelId)
+						}
+					}
+				}()
+			}
+		}
+
+		chanPref := model.Preference{
+			UserId:   c.Session.UserId,
+			Category: c.TeamId,
+			Name:     model.PREFERENCE_NAME_LAST_CHANNEL,
+			Value:    view.ChannelId,
+		}
+
+		teamPref := model.Preference{
+			UserId:   c.Session.UserId,
+			Category: model.PREFERENCE_CATEGORY_LAST,
+			Name:     model.PREFERENCE_NAME_LAST_TEAM,
+			Value:    c.TeamId,
+		}
+
+		Srv.Store.Preference().Save(&model.Preferences{teamPref, chanPref})
```

Comment:
We discussed and `BrowserStore` will be used. I will remove the preference writes from this PR.